components/dashboard/src/App.tsx

```diff
@@ -345,10 +378,77 @@ function App() {
         false,
     );
 
+    const acceptAllCookies = () => {
+        setCookie(true, true);
+        setCookiePreferences({
+            ...cookieConsentDefaultValue,
+            analytical: true,
+            targeting: true,
+        });
+        setShowCookieBanner(false);
+    };
+
+    const rejectAllCookies = () => {
+        setCookie(false, false);
+        setCookiePreferences({
+            ...cookieConsentDefaultValue,
+            analytical: false,
+            targeting: false,
+        });
+        setShowCookieBanner(false);
+    };
+
+    const openCookieSettings = () => {
+        setShowCookieSettings(true);
+    };
+
+    const handleAnalyticalCheckboxChange = () => {
+        setCookiePreferences({
+            ...cookiePreferences,
+            analytical: !cookiePreferences.analytical,
+        });
+    };
+
+    const handleTargetingCheckboxChange = () => {
+        setCookiePreferences({
+            ...cookiePreferences,
+            targeting: !cookiePreferences.targeting,
+        });
+    };
+
+    const saveCookiePreferences = () => {
+        setCookie(cookiePreferences.analytical, cookiePreferences.targeting);
+        setCookiePreferences({
+            ...cookieConsentDefaultValue,
+            analytical: cookiePreferences.analytical,
+            targeting: cookiePreferences.targeting,
+        });
+        setShowCookieSettings(false);
+        setShowCookieBanner(false);
+    };
+
+    const handleOnClose = () => {
+        const value = Cookies.get(`${window.location.hostname}-consent`);
+        if (value) {
+            setCookiePreferences(JSON.parse(value));
+        }
+        setShowCookieSettings(false);
+    };
+
     let toRender: React.ReactElement = (
         <Route>
             <div className="container">
                 <Menu />
+                {isGitpodIo() && showCookieSettings && (
```

Comment:
React question: is there a way to read this `showCookieSettings` from `CookieSettingsModal`? :thinking: 

because then we could push all the handlers into that modal, with it polluting `App.tsx` too much.

Reply:
It's true that App.tsx is already quite convoluted. But since the banner and settings both need to live in the App.tsx so that they will be shown across the app, and since both the settings and the banner need to inform the component that holds them both (App.tsx), then the App.tsx needs to be the one handling the state changes and reflecting the changing states[[1](https://reactjs.org/docs/lifting-state-up.html)]. 
I can sleep on it and think if other things can be extracted out..

Reply:
I extracted the non-shared states. It was a good suggestion 🙏🏽